File: cpp-scripts/remove-mismatching-rows.cpp
// remove_mismatching_rows.cpp
// Compile: g++ -std=c++17 -O3 -o remove_mismatching_rows remove_mismatching_rows.cpp
// Usage: ./remove_mismatching_rows input1.csv input2.csv output.csv --keys col1,col2

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <algorithm>
#include <filesystem>
#include <stdexcept>

namespace fs = std::filesystem;

// Custom hash function for vector<string> to use in unordered_set
struct VectorHash
{
    std::size_t operator()(const std::vector<std::string> &v) const
    {
        std::size_t seed = v.size();
        for (const auto &str : v)
        {
            seed ^= std::hash<std::string>{}(str) + 0x9e3779b9 + (seed << 6) + (seed >> 2);
        }
        return seed;
    }
};

class DataFrame
{
private:
    std::vector<std::string> columns_;
    std::vector<std::vector<std::string>> rows_;
    std::unordered_map<std::string, size_t> column_index_;

public:
    DataFrame() = default;

    // Read CSV file
    static DataFrame read_csv(const std::string &filepath)
    {
        DataFrame df;
        std::ifstream file(filepath);

        if (!file.is_open())
        {
            throw std::runtime_error("Cannot open file: " + filepath);
        }

        std::string line;
        bool first_row = true;

        while (std::getline(file, line))
        {
            // Remove carriage return if present
            if (!line.empty() && line.back() == '\r')
            {
                line.pop_back();
            }

            std::vector<std::string> row;
            std::stringstream ss(line);
            std::string cell;

            while (std::getline(ss, cell, ','))
            {
                // Basic CSV parsing (doesn't handle quoted fields with commas)
                // Trim whitespace
                cell.erase(0, cell.find_first_not_of(" \t"));
                cell.erase(cell.find_last_not_of(" \t") + 1);
                row.push_back(cell);
            }

            if (first_row)
            {
                df.columns_ = row;
                for (size_t i = 0; i < row.size(); ++i)
                {
                    df.column_index_[row[i]] = i;
                }
                first_row = false;
            }
            else
            {
                if (!row.empty())
                {
                    df.rows_.push_back(row);
                }
            }
        }

        return df;
    }

    // Write CSV file
    void write_csv(const std::string &filepath) const
    {
        // Create directory if it doesn't exist
        fs::path path(filepath);
        if (path.has_parent_path())
        {
            fs::create_directories(path.parent_path());
        }

        std::ofstream file(filepath);
        if (!file.is_open())
        {
            throw std::runtime_error("Cannot create output file: " + filepath);
        }

        // Write header
        for (size_t i = 0; i < columns_.size(); ++i)
        {
            file << columns_[i];
            if (i < columns_.size() - 1)
                file << ",";
        }
        file << "\n";

        // Write rows
        for (const auto &row : rows_)
        {
            for (size_t i = 0; i < row.size(); ++i)
            {
                file << row[i];
                if (i < row.size() - 1)
                    file << ",";
            }
            file << "\n";
        }
    }

    // Get column index
    size_t get_column_index(const std::string &col_name) const
    {
        auto it = column_index_.find(col_name);
        if (it == column_index_.end())
        {
            throw std::runtime_error("Column not found: " + col_name);
        }
        return it->second;
    }

    // Check if column exists
    bool has_column(const std::string &col_name) const
    {
        return column_index_.find(col_name) != column_index_.end();
    }

    // Get all column names
    const std::vector<std::string> &columns() const { return columns_; }

    // Get row count
    size_t row_count() const { return rows_.size(); }

    // Get column count
    size_t column_count() const { return columns_.size(); }

    // Get row data
    const std::vector<std::vector<std::string>> &rows() const { return rows_; }

    // Extract key values from a row
    std::vector<std::string> extract_keys(const std::vector<std::string> &row,
                                          const std::vector<size_t> &key_indices) const
    {
        std::vector<std::string> keys;
        keys.reserve(key_indices.size());
        for (size_t idx : key_indices)
        {
            if (idx < row.size())
            {
                keys.push_back(row[idx]);
            }
            else
            {
                keys.push_back("");
            }
        }
        return keys;
    }

    // Filter rows based on matching keys
    DataFrame filter_matching_rows(const DataFrame &reference,
                                   const std::vector<std::string> &key_columns) const
    {
        // Get key column indices for both dataframes
        std::vector<size_t> key_indices_this;
        std::vector<size_t> key_indices_ref;

        for (const auto &col : key_columns)
        {
            key_indices_this.push_back(this->get_column_index(col));
            key_indices_ref.push_back(reference.get_column_index(col));
        }

        // Build hash set of keys from reference dataframe
        std::unordered_set<std::vector<std::string>, VectorHash> reference_keys;
        for (const auto &row : reference.rows())
        {
            auto keys = reference.extract_keys(row, key_indices_ref);
            reference_keys.insert(std::move(keys));
        }

        // Filter this dataframe
        DataFrame result;
        result.columns_ = this->columns_;
        result.column_index_ = this->column_index_;

        for (const auto &row : this->rows_)
        {
            auto keys = this->extract_keys(row, key_indices_this);
            if (reference_keys.find(keys) != reference_keys.end())
            {
                result.rows_.push_back(row);
            }
        }

        return result;
    }
};

// Parse command line arguments
struct Arguments
{
    std::string input1;
    std::string input2;
    std::string output;
    std::vector<std::string> keys;
};

Arguments parse_arguments(int argc, char *argv[])
{
    Arguments args;

    if (argc < 5)
    {
        std::cerr << "Usage: " << argv[0] << " input1.csv input2.csv output.csv --keys col1,col2\n";
        std::cerr << "\nExamples:\n";
        std::cerr << "  " << argv[0] << " data1.csv data2.csv output.csv --keys id\n";
        std::cerr << "  " << argv[0] << " file1.csv file2.csv result.csv --keys name,email\n";
        throw std::runtime_error("Invalid arguments");
    }

    args.input1 = argv[1];
    args.input2 = argv[2];
    args.output = argv[3];

    // Find --keys argument
    for (int i = 4; i < argc; ++i)
    {
        std::string arg = argv[i];
        if (arg == "--keys" || arg == "-k")
        {
            if (i + 1 < argc)
            {
                std::string keys_str = argv[i + 1];
                std::stringstream ss(keys_str);
                std::string key;
                while (std::getline(ss, key, ','))
                {
                    // Trim whitespace
                    key.erase(0, key.find_first_not_of(" \t"));
                    key.erase(key.find_last_not_of(" \t") + 1);
                    if (!key.empty())
                    {
                        args.keys.push_back(key);
                    }
                }
                break;
            }
            else
            {
                throw std::runtime_error("--keys requires an argument");
            }
        }
    }

    if (args.keys.empty())
    {
        throw std::runtime_error("--keys argument is required");
    }

    return args;
}

// Validate columns exist in dataframe
void validate_columns(const DataFrame &df, const std::vector<std::string> &columns,
                      const std::string &filename)
{
    std::vector<std::string> missing;
    for (const auto &col : columns)
    {
        if (!df.has_column(col))
        {
            missing.push_back(col);
        }
    }

    if (!missing.empty())
    {
        std::cerr << "Error: Columns [";
        for (size_t i = 0; i < missing.size(); ++i)
        {
            std::cerr << missing[i];
            if (i < missing.size() - 1)
                std::cerr << ", ";
        }
        std::cerr << "] not found in " << filename << "\n";
        std::cerr << "Available columns: [";
        const auto &cols = df.columns();
        for (size_t i = 0; i < cols.size(); ++i)
        {
            std::cerr << cols[i];
            if (i < cols.size() - 1)
                std::cerr << ", ";
        }
        std::cerr << "]\n";
        throw std::runtime_error("Column validation failed");
    }
}

int main(int argc, char *argv[])
{
    try
    {
        // Parse arguments
        Arguments args = parse_arguments(argc, argv);

        std::cout << "Reading input files...\n";
        std::cout << "  Input 1 (reference): " << args.input1 << "\n";
        std::cout << "  Input 2 (comparison): " << args.input2 << "\n";
        std::cout << "  Primary key columns: [";
        for (size_t i = 0; i < args.keys.size(); ++i)
        {
            std::cout << args.keys[i];
            if (i < args.keys.size() - 1)
                std::cout << ", ";
        }
        std::cout << "]\n";

        // Read input files
        DataFrame df1 = DataFrame::read_csv(args.input1);
        DataFrame df2 = DataFrame::read_csv(args.input2);

        std::cout << "\nInput file statistics:\n";
        std::cout << "  " << args.input1 << ": " << df1.row_count()
                  << " rows, " << df1.column_count() << " columns\n";
        std::cout << "  " << args.input2 << ": " << df2.row_count()
                  << " rows, " << df2.column_count() << " columns\n";

        // Validate columns
        validate_columns(df1, args.keys, args.input1);
        validate_columns(df2, args.keys, args.input2);

        // Match rows
        std::cout << "\nMatching rows based on primary keys...\n";
        DataFrame matched_df = df1.filter_matching_rows(df2, args.keys);

        // Write output
        matched_df.write_csv(args.output);

        std::cout << "\nResults:\n";
        std::cout << "  Original rows in " << args.input1 << ": " << df1.row_count() << "\n";
        std::cout << "  Matching rows found: " << matched_df.row_count() << "\n";
        std::cout << "  Rows removed: " << (df1.row_count() - matched_df.row_count()) << "\n";
        std::cout << "  Output saved to: " << args.output << "\n";

        return 0;
    }
    catch (const std::exception &e)
    {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}

File: python-scripts/remove_mismatching_rows.py
#!/usr/bin/env python3
"""
Script to remove rows from the first document that don't match in the second document.
Supports both Excel and CSV files.

Usage:
    python script1.py input1.xlsx input2.csv output.xlsx --keys col1,col2
    python script1.py input1.csv input2.xlsx output.csv --keys id
"""

import pandas as pd
import argparse
import sys
from pathlib import Path


def read_file(filepath):
    """Read Excel or CSV file and return DataFrame"""
    filepath = Path(filepath)

    if not filepath.exists():
        raise FileNotFoundError(f"File not found: {filepath}")

    # Determine file type by extension
    extension = filepath.suffix.lower()

    if extension in [".xlsx", ".xls"]:
        return pd.read_excel(filepath)
    elif extension == ".csv":
        return pd.read_csv(filepath)
    else:
        raise ValueError(
            f"Unsupported file format: {extension}. Only .xlsx, .xls, and .csv are supported."
        )


def write_file(df, filepath):
    """Write DataFrame to Excel or CSV file"""
    filepath = Path(filepath)
    extension = filepath.suffix.lower()

    # Create output directory if it doesn't exist
    filepath.parent.mkdir(parents=True, exist_ok=True)

    if extension in [".xlsx", ".xls"]:
        df.to_excel(filepath, index=False)
    elif extension == ".csv":
        df.to_csv(filepath, index=False)
    else:
        raise ValueError(
            f"Unsupported output format: {extension}. Only .xlsx, .xls, and .csv are supported."
        )


def validate_columns(df, columns, filename):
    """Validate that specified columns exist in the DataFrame"""
    missing_cols = [col for col in columns if col not in df.columns]
    if missing_cols:
        print(f"Error: Columns {missing_cols} not found in {filename}")
        print(f"Available columns in {filename}: {list(df.columns)}")
        sys.exit(1)


def match_rows(df1, df2, key_columns):
    """
    Keep only rows from df1 that have matching key values in df2

    Args:
        df1: Reference DataFrame (rows will be removed from this)
        df2: Comparison DataFrame (used to find matches)
        key_columns: List of column names to use for matching

    Returns:
        DataFrame with only matching rows from df1
    """
    # Validate that key columns exist in both DataFrames
    validate_columns(df1, key_columns, "input file 1")
    validate_columns(df2, key_columns, "input file 2")

    # Create a set of tuples from df2 key columns for efficient lookup
    df2_keys = set()
    for _, row in df2.iterrows():
        key_tuple = tuple(row[col] for col in key_columns)
        df2_keys.add(key_tuple)

    # Filter df1 to keep only rows where key combination exists in df2
    def row_matches(row):
        key_tuple = tuple(row[col] for col in key_columns)
        return key_tuple in df2_keys

    matched_df = df1[df1.apply(row_matches, axis=1)].copy()

    return matched_df


def main():
    parser = argparse.ArgumentParser(
        description="Remove rows from first document that don't match in second document",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
    python script1.py data1.xlsx data2.csv output.xlsx --keys id
    python script1.py file1.csv file2.csv result.csv --keys name,email
    python script1.py input1.xlsx input2.xlsx output.csv --keys user_id,date
        """,
    )

    parser.add_argument(
        "input1", help="First input file (Excel or CSV) - reference document"
    )
    parser.add_argument(
        "input2", help="Second input file (Excel or CSV) - comparison document"
    )
    parser.add_argument("output", help="Output file (Excel or CSV)")
    parser.add_argument(
        "--keys",
        "-k",
        help="Comma-separated list of column names to use as primary keys for matching",
        required=True,
    )

    args = parser.parse_args()

    try:
        # Parse key columns
        key_columns = [col.strip() for col in args.keys.split(",")]

        print(f"Reading input files...")
        print(f"  Input 1 (reference): {args.input1}")
        print(f"  Input 2 (comparison): {args.input2}")
        print(f"  Primary key columns: {key_columns}")

        # Read input files
        df1 = read_file(args.input1)
        df2 = read_file(args.input2)

        print(f"\nInput file statistics:")
        print(f"  {args.input1}: {len(df1)} rows, {len(df1.columns)} columns")
        print(f"  {args.input2}: {len(df2)} rows, {len(df2.columns)} columns")

        # Match rows
        print(f"\nMatching rows based on primary keys...")
        matched_df = match_rows(df1, df2, key_columns)

        # Write output
        write_file(matched_df, args.output)

        print(f"\nResults:")
        print(f"  Original rows in {args.input1}: {len(df1)}")
        print(f"  Matching rows found: {len(matched_df)}")
        print(f"  Rows removed: {len(df1) - len(matched_df)}")
        print(f"  Output saved to: {args.output}")

    except Exception as e:
        print(f"Error: {e}", file=sys.stderr)
        sys.exit(1)


if __name__ == "__main__":
    main()
